warehouse/query-core/src/main/java/datawave/query/transformer/GroupingTransform.java

```diff
@@ -93,81 +163,185 @@ private Key createAttrKey(String fieldName) {
+                }
             }
         };
-        
     }
     
     @Override
     public Entry<Key,Document> flush() {
+        if (flatten)
+            log.trace("tserver flush");
+        else
+            log.trace("webserver flush");
         if (documents == null) {
             documents = new LinkedList<>();
         }
-        if (!multiset.isEmpty()) {
-            if (log.isTraceEnabled()) {
-                log.trace("flush will use the multiset:" + multiset);
-            }
-            for (Collection<Attribute<?>> entry : multiset.elementSet()) {
-                ColumnVisibility vis;
+        if (!countingMap.isEmpty()) {
+            
+            log.trace("flush will use the countingMap: {}", countingMap);
+            
+            for (Collection<GroupingTypeAttribute<?>> entry : countingMap.keySet()) {
+                log.trace("from countingMap, got entry: {}", entry);
+                ColumnVisibility columnVisibility = null;
                 try {
-                    vis = toColumnVisibility(fieldVisibilities.get(entry));
+                    columnVisibility = toColumnVisibility(fieldVisibilities.get(entry));
                 } catch (Exception e) {
                     throw new IllegalStateException("Unable to merge column visibilities: " + fieldVisibilities.get(entry), e);
                 }
                 // grab a key from those saved during getListKeyCounts
                 Assert.notEmpty(keys, "no available keys for grouping results");
-                Key docKey = keys.get(0);
+                Key docKey = keys.get(keys.size() - 1);
                 Document d = new Document(docKey, true);
+                d.setColumnVisibility(columnVisibility);
                 
-                for (Attribute base : entry) {
-                    d.put(getFieldName(base), base);
-                }
+                entry.forEach(base -> d.put(getFieldName(base), base));
                 NumberType type = new NumberType();
-                type.setDelegate(new BigDecimal(multiset.count(entry)));
+                type.setDelegate(new BigDecimal(countingMap.get(entry)));
                 TypeAttribute<BigDecimal> attr = new TypeAttribute<>(type, new Key("count"), true);
                 d.put("COUNT", attr);
                 documents.add(d);
             }
+            if (flatten) {
+                // flatten to just one document on the tserver.
+                flatten(documents);
+            }
         }
         if (!documents.isEmpty()) {
+            log.trace("{} will flush first of {} documents: {}", this.hashCode(), documents.size(), documents);
             Document d = documents.pop();
-            Entry<Key,Document> entry = Maps.immutableEntry(d.getMetadata(), d);
-            if (log.isTraceEnabled()) {
-                log.trace("flushing out " + entry);
+            Key key;
+            if (keys.size() > 0 && flatten) {
+                key = keys.get(keys.size() - 1);
```

Comment:
A comment as to why you selected the last one. I presume because in the end you need the last key looked at to be the one returned for teardown/rebuilt purposes.